alf/utils/adaptive_normalizer.py

```diff
@@ -14,73 +14,97 @@
 
 import tensorflow as tf
 
-from tf_agents.utils.tensor_normalizer import EMATensorNormalizer
 from tf_agents.specs.tensor_spec import TensorSpec
+from alf.utils.averager import AdaptiveAverager
 
 
-class AdaptiveNormalizer(EMATensorNormalizer):
+class AdaptiveNormalizer(tf.Module):
     def __init__(self,
                  tensor_spec,
                  speed=2.0,
                  auto_update=True,
-                 variance_epsilon=1e-3):
-        """Create reward normalizer
+                 variance_epsilon=1e-8,
+                 name="AdaptiveNormalizer"):
+        """Create an adaptive normalizer.
 
         This normalizer gives higher weight to more recent samples for
         calculating mean and variance. Roughly speaking, the weight for each
         sample at time t is proportional to (t/T)^(speed-1), where T is the
         current time step. See docs/streaming_averaging_amd_sampling.py for
         detail.
 
+        Given weights w_i and samples x_i, i = 1...n, let
+
+        m   = \sum_i w_i * x_i
+        m2  = \sum_i w_i * x_i^2
+
+        then
+
+        var = \sum_i w_i * (x_i - m)^2
+            = \sum_i w_i * (x_i^2 + m^2 - 2*x_i*m)
+            = m2 + m^2 - 2m^2
+            = m2 - m^2
+
+        which is the same result with the case when w_1=w_2=...=w_n=(1/n)
+
         Args:
             tensor_spec (TensorSpec): spec of the mean of tensors to be
               normlized.
             speed (float): speed of updating mean and variance.
             auto_update (bool): If True, automatically update mean and variance
               for each call to `normalize()`. Otherwise, the user needs to call
               `update()`
+            variance_epislon (float): a small value added to std for normalizing
+            name (str):
         """
-        self._total_env_steps = tf.Variable(
-            int(speed), dtype=tf.int64, trainable=False)
-        self._update_ema_rate = tf.Variable(
-            1.0, dtype=tf.float32, trainable=False)
-        self._speed = speed
+        self._mean_averager = AdaptiveAverager(tensor_spec, speed=speed)
+        self._m2_averager = AdaptiveAverager(tensor_spec, speed=speed)
         self._auto_update = auto_update
         self._variance_epsilon = variance_epsilon
-        super(AdaptiveNormalizer, self).__init__(
-            tensor_spec, norm_update_rate=self._update_ema_rate)
 
-    def normalize(self, tensors, clip_value=-1.0):
-        """Normalized the reward
+    def update(self, tensor):
```

Comment:
you need to first get the outer rank, and reduce over the outer ranks. tensor may or may not be batched.
```python
tf_agent.utils.nest_utils.get_outer_rank(tensor, tensor_spec)
```